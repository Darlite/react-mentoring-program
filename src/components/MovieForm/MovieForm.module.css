.movieForm {
    display: grid;
    gap: 30px;
    grid-template-columns: 2fr 1fr;
}

.inputLabel {
    text-transform: uppercase;
    color: #F65261;
    font-size: 16px;
    font-weight: 600;
    letter-spacing: 0.89px;
}

.input {
    background: rgba(50, 50, 50, 0.95);
    color: #FFFFFF;
    padding: 15px;
    font-size: 20px;
    border: none;
    border-radius: 4px;
}

.selectInput {
    color: #000000;
    font-size: 20px;
    border: none;
    border-radius: 4px;
}

.resetButton {
    border: 1px solid #F65261;
    font-size: 20px;
    color: #F65261;
    text-transform: uppercase;
    background: #232323;
    padding: 16px 35px;
    cursor: pointer;
}

.submitButton {
    background-color: #F65261;
    font-size: 20px;
    color: #FFF;
    text-transform: uppercase;
    border: 1px solid #F65261;
    padding: 16px 35px;
    cursor: pointer;
}

.labelAndInputContainer {
    display: flex;
    row-gap: 10px;
    flex-direction: column;
}

.overviewContainer {
    display: grid;
    row-gap: 10px;
    grid-column-start: 1;
    grid-column-end: 3;
}

.overviewContainer textarea {
    min-height: 150px;
}

.buttonsContainer {
    display: flex;
    gap: 15px;
    justify-content: end;
    grid-column-start: 2;
}
