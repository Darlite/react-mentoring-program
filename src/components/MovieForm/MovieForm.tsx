import styles from "../MovieForm/MovieForm.module.css";
import React, {FormEvent} from "react";
import {MovieDetailsData} from "../../types/MovieDetailsData";
import {GenreType} from "../../constants/GenreType";
import InputField from "../InputField/InputField";
import GenreSelect from "../GenreSelect/GenreSelect";

interface MovieFormProps {
    initialMovieInfo?: MovieDetailsData | null,
    handleSubmit: (movieData: MovieDetailsData) => void;
}

const MovieForm: React.FC<MovieFormProps> = ({ initialMovieInfo, handleSubmit }) => {
    function onSubmit(e: FormEvent<HTMLFormElement>) {
        e.preventDefault();
        const form = e.target as HTMLFormElement;
        const formData = new FormData(form);
        const formDataObject = Object.fromEntries(formData) as Record<string, string>;

        const genres = formData.getAll("genres") as string[] as GenreType[];

        const id = initialMovieInfo?.id ?? Math.floor(Math.random() * 10000);

        const movieData = {
            id,
            poster_path: initialMovieInfo?.poster_path || "",
            title: formDataObject.title,
            release_date: formDataObject.release_date,
            movieUrl: formDataObject.movieUrl,
            vote_average: formDataObject.vote_average,
            genres,
            runtime: formDataObject.runtime,
            overview: formDataObject.overview,
        };

        handleSubmit(movieData);
    }

    return (
        <form className={styles.movieForm}
              onSubmit={onSubmit}
        >
            <InputField id="movieTitle"
                        label="Title"
                        type="text"
                        name="title"
                        placeholder="Movie title"
                        defaultValue={initialMovieInfo?.title || ""}
                        required={true}
                        ariaLabel="Movie title"
            />

            <InputField id="releaseDate"
                        label="Release Date"
                        type="date"
                        name="releaseDate"
                        placeholder="Select Date"
                        defaultValue={initialMovieInfo?.release_date || ""}
                        required={true}
                        ariaLabel="Release Date"
            />

            <InputField id="movieUrl"
                        label="Movie URL"
                        type="url"
                        name="movieUrl"
                        placeholder="https://"
                        defaultValue={initialMovieInfo?.movieUrl || ""}
                        required={true}
                        ariaLabel="Movie URL"
            />

            <InputField id="rating"
                        label="Rating"
                        type="number"
                        name="rating"
                        placeholder="7.8"
                        defaultValue={initialMovieInfo?.vote_average || ""}
                        required={true}
                        ariaLabel="Movie Rating"
            />

            <GenreSelect defaultOptions={initialMovieInfo?.genres} />

            <InputField id="runtime"
                        label="Runtime"
                        type="text"
                        name="runtime"
                        placeholder="minutes"
                        defaultValue={initialMovieInfo?.runtime || ""}
                        required={true}
                        ariaLabel="Movie Runtime"
            />

            <div className={styles.overviewContainer}>
                <label className={styles.inputLabel}
                       htmlFor="overview">Overview</label>
                <textarea className={styles.input}
                          id="overview"
                          name="overview"
                          aria-label="Movie overview"
                          defaultValue={initialMovieInfo?.overview || ""}
                          placeholder="Movie overview"
                />
            </div>

            <div className={styles.buttonsContainer}>
                <button className={styles.resetButton}
                        type="reset"
                        aria-label="Reset"
                >
                    Reset
                </button>
                <button className={styles.submitButton}
                        type="submit"
                        aria-label="Submit"
                >
                    Submit
                </button>
            </div>
        </form>
    )
}

export default MovieForm;
