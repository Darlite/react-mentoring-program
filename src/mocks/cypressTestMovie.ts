import {GenreType} from "../constants/GenreType";

export const cypressTestMovie = {
    title: "Cypress Test Movie",
    editedMovieTitle: "Cypress Edited Test Movie",
    release_date: "2025-05-04",
    poster_path: "https://www.movieposters.com/cdn/shop/files/tron-ares_avxg7mih_480x.progressive.jpg",
    genres: [GenreType.Action, GenreType.ScienceFiction],
    runtime: "138",
    vote_average: "7.2",
    overview: "A highly sophisticated Program, Ares, is sent from the digital world into the real world on a dangerous mission.",
}
