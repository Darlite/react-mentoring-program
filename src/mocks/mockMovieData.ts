import {GenreType} from "../constants/GenreType";

export const moviesData = {
    id: 1,
    imageUrl: "images/Pulp_Fiction.png",
    title: "Pulp Fiction",
    releaseDate: "1994-09-23",
    movieUrl: "https://www.imdb.com/title/tt0110912",
    genres: [GenreType.Crime, GenreType.Thriller],
    runtime: "2h 34m",
    rating: "8.9",
    description: "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
}
