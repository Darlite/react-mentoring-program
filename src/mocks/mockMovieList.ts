import {GenreType} from "../constants/GenreType";

export const mockMovieList = [
    {
        id: 1,
        poster_path: "images/Pulp_Fiction.png",
        title: "Pulp Fiction",
        release_date: "1994-09-23",
        movieUrl: "https://www.imdb.com/title/tt0110912",
        genres: [GenreType.Crime, GenreType.Thriller],
        runtime: "2h 34m",
        vote_average: "8.9",
        overview: "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
    },
    {
        id: 2,
        poster_path: "images/Bohemian_Rhapsody.png",
        title: "Bohemian Rhapsody",
        release_date: "2018-10-30",
        movieUrl: "https://www.imdb.com/title/tt1727824",
        genres: [GenreType.Musical, GenreType.Documentary],
        runtime: "2h 15m",
        vote_average: "8.5",
        overview: "With his impeccable vocal abilities, Freddie Mercury and his rock band, Queen, achieve superstardom. However, amidst his skyrocketing success, he grapples with his ego, sexuality and a fatal illness.",
    },
    {
        id: 3,
        poster_path: "images/Kill_Bill_Volume_2.png",
        title: "Kill Bill: Vol. 2",
        release_date: "2004-04-08",
        movieUrl: "https://www.imdb.com/title/tt0378194",
        genres: [GenreType.Action, GenreType.Thriller],
        runtime: "2h 17m",
        vote_average: "8.9",
        overview: "A pregnant woman, codenamed the Bride, sets out on a journey to kill her ex-boss, Bill, and targets his brother, Budd, and Elle Driver, the only two survivors of the Deadly Vipers Assassination Squad.",
    },
]