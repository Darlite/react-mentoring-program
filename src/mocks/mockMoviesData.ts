import {GenreType} from "../constants/GenreType";

export const mockMoviesData = {
    id: 680,
    poster_path: "https://www.movieposters.com/cdn/shop/files/PulpFiction.Ver1_480x.progressive.jpg?v=1741192018",
    title: "Pulp Fiction",
    release_date: "1994-09-10",
    genres: [GenreType.Crime, GenreType.Thriller],
    runtime: "167",
    vote_average: "8.3",
    overview: "Jules Winnfield (Samuel L. Jackson) and Vincent Vega (John Travolta) are two hit men who are out to retrieve a suitcase stolen from their employer, mob boss Marsellus Wallace (Ving Rhames). Wallace has also asked Vincent to take his wife Mia (Uma Thurman) out a few days later when Wallace himself will be out of town. Butch Coolidge (Bruce Willis) is an aging boxer who is paid by Wallace to lose his fight. The lives of these seemingly unrelated people are woven together comprising of a series of funny, bizarre and uncalled-for incidents.â€”Soumitra",
}
