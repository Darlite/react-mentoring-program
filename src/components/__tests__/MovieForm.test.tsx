import MovieForm from "../MovieForm/MovieForm";
import React from "react";
import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {moviesData} from "../../mocks/mockMovieData";

describe("MovieForm component", () => {
    it("renders correctly", () => {
        const { container } = render(<MovieForm handleSubmit={() => {}}
        />);
        expect(container).toMatchSnapshot();
    });

    it("renders with initial movie data", () => {
        render(<MovieForm initialMovieInfo={moviesData}
                          handleSubmit={() => {}}
        />);

        const inputTitle = screen.getByRole("textbox", { name: /movie title/i });
        expect(inputTitle).toHaveValue(moviesData.title);

        const inputReleaseDate = screen.getByLabelText(/release date/i);
        expect(inputReleaseDate).toHaveValue(moviesData.releaseDate);

        const inputMovieURL = screen.getByRole("textbox", { name: /movie url/i });
        expect(inputMovieURL).toHaveValue(moviesData.movieUrl);

        const inputMovieRating = screen.getByRole("spinbutton", { name: /movie rating/i });
        expect(inputMovieRating).toHaveValue(parseFloat(moviesData.rating));

        const selectGenres = screen.getAllByRole("option", { selected: true });
        const selectedGenres = selectGenres.map((option) => (option as HTMLOptionElement).value);
        expect(selectedGenres).toEqual(moviesData.genres);

        const inputRuntime = screen.getByRole("textbox", { name: /movie runtime/i });
        expect(inputRuntime).toHaveValue(moviesData.runtime);

        const inputDescription = screen.getByRole("textbox", { name: /movie description/i });
        expect(inputDescription).toHaveValue(moviesData.description);
    });
})
