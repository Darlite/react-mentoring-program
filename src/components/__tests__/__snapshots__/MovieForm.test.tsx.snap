// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm component renders correctly 1`] = `
<div>
  <form
    class="movieForm"
  >
    <div
      class="labelAndInputContainer"
    >
      <label
        class="inputLabel"
        for="movieTitle"
      >
        Title
      </label>
      <input
        aria-label="Movie title"
        class="input"
        id="movieTitle"
        name="title"
        placeholder="Movie title"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="labelAndInputContainer"
    >
      <label
        class="inputLabel"
        for="releaseDate"
      >
        Release Date
      </label>
      <input
        aria-label="Release Date"
        class="input"
        id="releaseDate"
        name="releaseDate"
        placeholder="Select Date"
        required=""
        type="date"
        value=""
      />
    </div>
    <div
      class="labelAndInputContainer"
    >
      <label
        class="inputLabel"
        for="movieUrl"
      >
        Movie URL
      </label>
      <input
        aria-label="Movie URL"
        class="input"
        id="movieUrl"
        name="movieUrl"
        placeholder="https://"
        required=""
        type="url"
        value=""
      />
    </div>
    <div
      class="labelAndInputContainer"
    >
      <label
        class="inputLabel"
        for="rating"
      >
        Rating
      </label>
      <input
        aria-label="Movie Rating"
        class="input"
        id="rating"
        name="rating"
        placeholder="7.8"
        required=""
        type="number"
        value=""
      />
    </div>
    <div
      class="labelAndInputContainer"
    >
      <label
        class="inputLabel"
        for="genres"
      >
        Genre
      </label>
      <select
        aria-label="Movie Genres"
        class="input"
        id="genres"
        multiple=""
        name="genres"
        required=""
      >
        <option
          value="Action"
        >
          Action
        </option>
        <option
          value="Crime"
        >
          Crime
        </option>
        <option
          value="Documentary"
        >
          Documentary
        </option>
        <option
          value="Thriller"
        >
          Thriller
        </option>
        <option
          value="Musical"
        >
          Musical
        </option>
      </select>
    </div>
    <div
      class="labelAndInputContainer"
    >
      <label
        class="inputLabel"
        for="runtime"
      >
        Runtime
      </label>
      <input
        aria-label="Movie Runtime"
        class="input"
        id="runtime"
        name="runtime"
        placeholder="minutes"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="overviewContainer"
    >
      <label
        class="inputLabel"
        for="description"
      >
        Overview
      </label>
      <textarea
        aria-label="Movie Description"
        class="input"
        id="description"
        name="description"
        placeholder="Movie description"
      />
    </div>
    <div
      class="buttonsContainer"
    >
      <button
        class="resetButton"
        type="reset"
      >
        Reset
      </button>
      <button
        class="submitButton"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
